File: Controler/SideControl/SideControl.py
# Constant Addresses
TOP_PANEL = 0x50
RIGHT_PANEL = 0x51
BOTTOM_PANEL = 0x52
LEFT_PANEL = 0x53

class SideController:
    import smbus
    clsbusctrl = None
    def __init__(busctrl: smbus.SMBus):
        clsbusctrl = busctrl
    
    # Resets a indicators value to nothing
    def clearIndicator(address: int, indicatornum: int):
        if address not in range(256) or indicatornum not in range(2):
            return -1
        # write data to the specified side panel
        # first byte is thje indicator number, second byte is if it is lit, following 3 bytes are char
        # 0x11 is set indictator
        # the first 4 bits of byte 3 are the indicator adddress and bits 567 are not used and bit 8 is whether or not the indicator is lit
        clsbusctrl.write_i2c_block_data(address, 0x11, indicatornum << 4, "")
        
    def setIndicator(address: int, indicatornum: int, indicatortext: str, lit: bool):
        if address not in range(256) or indicatornum not in range(2) or len(str) > 3:
            return -1
        # write data to the specified side panel
        # first byte is thje indicator number, second byte is if it is lit, following 3 bytes are char
        # 0x11 is set indictator
        # the first 4 bits of byte 3 are the indicator adddress and bits 567 are not used and bit 8 is whether or not the indicator is lit
        indicatorNumlit = (indicatornum << 4) + lit
        clsbusctrl.write_i2c_block_data(address, 0x11, indicatorNumlit, indicatortext)

    def clearSerialNumber():
        clsbusctrl.write_i2c_block_data(BOTTOM_PANEL, 0x12, "")

    def setSerialNumber(SerialNumber: str):
        if SerialNumber > 6:
            return -1
        # 0x12 is set Serial Number
        clsbusctrl.write_i2c_block_data(BOTTOM_PANEL, 0x12, SerialNumber)

File: Controler/SideControl/SideControl.cpp
#include "SideControl.h"

SideControl::SideControl(int *busname)
{
    SideControl::file_i2c = &busname;
}
SideControl::setSerialNumber(char serialnumber[8]) {
    uint8_t buffer[9];
    buffer[0] = 0x10;
    bytestosend = 1;
    for (int i = 0;i < 8; i++){
        if (serialNumber[i] != 0){
            buffer[i+1] = serialNumber[i];
            bytestosend++;
        }
    }   
    if (ioctl(SideControl::file_i2c, I2C_SLAVE, SideControl::RIGHT_PANEL)) {
        printf("Failed to acquire bus access and/or talk to slave.\n");
        return -1;
    }
    if (write(SideControl::file_i2c, buffer, bytestosend) != bytestosend) {
        printf("Failed to write to the i2c bus.\n");
        return -1;
    }
    return 1;
}
SideControl::clearSerialNumber() {
    if (ioctl(SideControl::file_i2c, I2C_SLAVE, SideControl::RIGHT_PANEL)) {
        printf("Failed to acquire bus access and/or talk to slave.\n");
        return -1;
    }
    if (write(SideControl::file_i2c, {0x20}, 1) != 1) {
        printf("Failed to write to the i2c bus.\n");
        return -1;
    }
    return 1;
}
SideControl::setIndicator(uint8_t panel, bool lit, char indicator[3]) {\
    uint8_t buffer[5];
    buffer[0] = 0x11;
    buffer[1] = lit;
    for (int i = 0;i < 3; i++){
        buffer[i+2] = indicator[i];
    }
    if (ioctl(SideControl::file_i2c, I2C_SLAVE, panel)) {
        printf("Failed to acquire bus access and/or talk to slave.\n");
        return -1;
    }
    if (write(SideControl::file_i2c, buffer, 5) != 5) {
        printf("Failed to write to the i2c bus.\n");
        return -1;
    }
    return 1;
}
SideControl::clearAllIndicator(uint8_t panel) {
    if (ioctl(SideControl::file_i2c, I2C_SLAVE, panel)) {
        printf("Failed to acquire bus access and/or talk to slave.\n");
        return -1;
    }
    if (write(SideControl::file_i2c, {0x21}, 1) != 1) {
        printf("Failed to write to the i2c bus.\n");
        return -1;
    }
    return 1;
}
SideControl::setSideArt(uint8_t panel, uint8_t art]) {
    if (ioctl(SideControl::file_i2c, I2C_SLAVE, panel)) {
        printf("Failed to acquire bus access and/or talk to slave.\n");
        return -1;
    }
    if (write(SideControl::file_i2c, {0x12, art}, 2) != 2) {
        printf("Failed to write to the i2c bus.\n");
        return -1;
    }
    return 1;
}
SideControl::clearAllSideArt(uint8_t panel) {
    if (ioctl(SideControl::file_i2c, I2C_SLAVE, panel)) {
        printf("Failed to acquire bus access and/or talk to slave.\n");
        return -1;
    }
    if (write(SideControl::file_i2c, {0x22}, 1) != 1) {
        printf("Failed to write to the i2c bus.\n");
        return -1;
    }
    return 1;
}